# Exercices XP Gold
Dernière mise à jour : 20 novembre 2024

👩‍🏫 👩🏿‍🏫 Ce que vous apprendrez
Techniques avancées dans NumPy pour la gestion des tableaux aléatoires, la normalisation des matrices et l'espacement des tableaux.
Compétences pour trouver le minimum, le maximum et effectuer une normalisation sur des matrices.


🛠️ Ce que vous allez créer
Tableaux NumPy avancés, comprenant un tableau 5x5 avec des valeurs aléatoires, une matrice 3x3 normalisée et des éléments régulièrement espacés dans un tableau.


Brève introduction aux matrices
Qu'est-ce qu'une matrice en Python et NumPy ?

Une matrice est un tableau bidimensionnel de nombres disposés en lignes et en colonnes. En Python, les matrices peuvent être représentées par des listes de listes ou, plus efficacement, par des tableaux NumPy.
NumPy fournit un moyen puissant et efficace de travailler avec des matrices, offrant diverses fonctions pour les opérations matricielles.
Exemple de matrice :



# Example of a 3x3 matrix using NumPy
import numpy as np
matrix_example = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
print(matrix_example)


## Exercice 1 : Minimum et maximum d'un tableau aléatoire
Instructions
Créez un tableau 5x5 avec des valeurs aléatoires et trouvez les valeurs minimales et maximales.

Résultat attendu :



Min: 0.01, Max: 0.99

In [2]:
import numpy as np  # Importation de NumPy

In [3]:
# Étape 1 : Créer un tableau 5x5 avec des valeurs aléatoires entre 0 et 1
tableau = np.random.rand(5, 5)

In [4]:
# Étape 2 : Trouver les valeurs minimale et maximale du tableau
val_min = np.min(tableau)
val_max = np.max(tableau)

In [5]:
# Étape 3 : Arrondir les résultats pour lisibilité
val_min = round(val_min, 2)
val_max = round(val_max, 2)

In [6]:
# Étape 4 : Affichage des résultats
print("Min:", val_min, ", Max:", val_max)

Min: 0.01 , Max: 0.98


## Exercice 2 : Normalisation de matrice
Instructions
Normaliser une matrice aléatoire 3x3 (soustraire la moyenne et diviser par l'écart type de la matrice).

Résultat attendu :



array([[-1.22474487,  1.22474487],
       [ 0.        ,  0.        ]])

In [7]:
import numpy as np  # Importation de NumPy

In [8]:
# Étape 1 : Créer une matrice 3x3 avec des valeurs aléatoires
matrice = np.random.rand(3, 3)

In [9]:
# Étape 2 : Calcul de la moyenne et de l'écart type
moyenne = np.mean(matrice)
ecart_type = np.std(matrice)

In [10]:
# Étape 3 : Normalisation de la matrice
# Chaque valeur devient : (valeur - moyenne) / écart-type
matrice_normalisee = (matrice - moyenne) / ecart_type

In [11]:
# Étape 4 : Affichage du résultat
print(matrice_normalisee)

[[ 0.40854559 -0.99369978  0.23285991]
 [-0.96976932 -0.21803729  1.28313249]
 [-1.6608591   1.48783795  0.42998955]]


## Exercice 3 : Éléments régulièrement espacés dans un tableau
Instructions
Créez un tableau 1D de 50 éléments régulièrement espacés entre 0 et 10, exclusifs.

Résultat attendu :



array([0. , 0.2, 0.4, ..., 9.6, 9.8, 10.])

In [12]:
import numpy as np  # Importation de NumPy

In [13]:
# Étape 1 : Créer un tableau 1D de 50 valeurs espacées entre 0 et 10
# np.linspace génère des valeurs espacées de façon linéaire
tableau = np.linspace(0, 10, 50)

In [14]:
# Étape 2 : Affichage du tableau
print(tableau)

[ 0.          0.20408163  0.40816327  0.6122449   0.81632653  1.02040816
  1.2244898   1.42857143  1.63265306  1.83673469  2.04081633  2.24489796
  2.44897959  2.65306122  2.85714286  3.06122449  3.26530612  3.46938776
  3.67346939  3.87755102  4.08163265  4.28571429  4.48979592  4.69387755
  4.89795918  5.10204082  5.30612245  5.51020408  5.71428571  5.91836735
  6.12244898  6.32653061  6.53061224  6.73469388  6.93877551  7.14285714
  7.34693878  7.55102041  7.75510204  7.95918367  8.16326531  8.36734694
  8.57142857  8.7755102   8.97959184  9.18367347  9.3877551   9.59183673
  9.79591837 10.        ]


## Exercice 4 : Multiplication de matrices
Instructions
Multiplier une matrice 5x3 par une matrice 3x2 (produit matriciel réel).

Résultat attendu :



array([[ 8.79,  6.73],
       [12.87,  9.77],
       [16.92, 12.83],
       [20.98, 15.89],
       [25.04, 18.95]])

In [15]:
import numpy as np  # Importation de NumPy

In [16]:
# Étape 1 : Définir une matrice 5x3
A = np.array([
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9],
    [10, 11, 12],
    [13, 14, 15]
])

In [17]:
# Étape 2 : Définir une matrice 3x2
B = np.array([
    [0.5, 0.2],
    [1.0, 0.4],
    [1.5, 0.6]
])

In [18]:
# Étape 3 : Effectuer le produit matriciel
# Utiliser np.dot() ou l'opérateur @
resultat = A.dot(B)  # ou A @ B

In [19]:
# Étape 4 : Affichage du résultat
print(np.round(resultat, 2))

[[ 7.   2.8]
 [16.   6.4]
 [25.  10. ]
 [34.  13.6]
 [43.  17.2]]


## Exercice 5
Instructions
Utilisez votre compréhension de la multiplication matricielle pour résoudre cet exercice.
Créez deux matrices de dimensions compatibles.
Effectuez une multiplication matricielle sur ces matrices et affichez le résultat.
Assurez-vous que les dimensions des matrices permettent une multiplication valide.
Conseil : n'oubliez pas que le nombre de colonnes de la première matrice doit correspondre au nombre de lignes de la deuxième matrice pour une multiplication réussie.

In [20]:
import numpy as np  # Import de NumPy

In [21]:
# Étape 1 : Créer une matrice A de dimension 3x4
A = np.array([
    [1, 2, 3, 4],
    [5, 6, 7, 8],
    [9, 10, 11, 12]
])

In [22]:
# Étape 2 : Créer une matrice B de dimension 4x2 (compatible avec A)
B = np.array([
    [2, 0],
    [1, 2],
    [0, 1],
    [3, 4]
])


In [23]:
# Étape 3 : Effectuer la multiplication matricielle A x B
resultat = A.dot(B)  # ou A @ B

In [24]:
# Étape 4 : Afficher le résultat
print(resultat)

[[16 23]
 [40 51]
 [64 79]]
